File: Stripe/STPPaymentMethodCardNetworks.m
//
//  STPPaymentMethodCardNetworks.m
//  Stripe
//
//  Created by Cameron Sabol on 7/15/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

#import "STPPaymentMethodCardNetworks.h"

#import "NSDictionary+Stripe.h"

NS_ASSUME_NONNULL_BEGIN

@implementation STPPaymentMethodCardNetworks

@synthesize allResponseFields = _allResponseFields;

- (NSString *)description {
    NSArray *props = @[
                       // Object
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],

                       // Properties
                       [NSString stringWithFormat:@"available: %@", self.available],
                       [NSString stringWithFormat:@"preferred: %@", self.preferred],
                       ];
    
    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPAPIResponseDecodable

- (nullable instancetype)_initWithDictionary:(nullable NSDictionary *)dict {
    dict = [dict stp_dictionaryByRemovingNulls];
    if (!dict) {
        return nil;
    }
    NSArray<NSString *> *available = [dict stp_arrayForKey:@"available" withObjectType:[NSString class]];
    if (available == nil) {
        return nil;
    }
        
    self = [super init];
    if (self) {
        self->_allResponseFields = [dict copy];
        self->_available = available;
        self->_preferred = [[dict stp_stringForKey:@"preferred"] copy];
    }
    
    return self;
}

+ (nullable instancetype)decodedObjectFromAPIResponse:(nullable NSDictionary *)response {
    return [[self alloc] _initWithDictionary:response];
}
                       
@end

NS_ASSUME_NONNULL_END

File: Stripe/STPPaymentMethodCardNetworks.swift
//
//  STPPaymentMethodCardNetworks.swift
//  Stripe
//
//  Created by Cameron Sabol on 7/15/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

import Foundation

/// `STPPaymentMethodCardNetworks` contains information about card networks that can be used to process a payment.
public class STPPaymentMethodCardNetworks: NSObject, STPAPIResponseDecodable {
  public var allResponseFields: [AnyHashable: Any] = [:]

  /// All available networks for the card.
  public private(set) var available: [String] = []
  /// The preferred network for the card if one exists.
  public private(set) var preferred: String?

  /// :nodoc:
  @objc public override var description: String {
    let props = [
      // Object
      String(format: "%@: %p", NSStringFromClass(STPPaymentMethodCardNetworks.self), self),
      // Properties
      "available: \(available)",
      "preferred: \(preferred ?? "")",
    ]

    return "<\(props.joined(separator: "; "))>"
  }

  // MARK: - STPAPIResponseDecodable
  required init?(withDictionary dict: [AnyHashable: Any]) {
    super.init()
    let nsDict = (dict as NSDictionary).stp_dictionaryByRemovingNulls() as NSDictionary
    guard let available = nsDict.stp_array(forKey: "available") as? [String] else {
      return nil
    }
    self.available = available
    self.preferred = nsDict.stp_string(forKey: "preferred")
    self.allResponseFields = dict
  }

  public class func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
    guard let response = response else {
      return nil
    }
    return self.init(withDictionary: response)
  }
}
